File: C++/33.传输对象模式/TransferObjectPatternDemo.cpp
/**
 * @file TransferObjectPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-12-11
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "StudentBO.h"

int main(int argc, char const *argv[]) {
    std::shared_ptr<StudentBO> student_business_object = std::make_shared<StudentBO>();

    std::shared_ptr<StudentVO> student = nullptr;

    //输出所有的学生
    for (std::shared_ptr<StudentVO> student : (*student_business_object->GetAllStudents())) {
        std::cout << "Student: [RollNo : " << student->GetRollNo() << ", Name : " << student->GetName() << " ]"
                  << std::endl;
    }

    //更新学生
    student = student_business_object->GetAllStudents()->front();
    student->SetName("Michael");
    student_business_object->UpdateStudent(student);

    //获取学生
    student = student_business_object->GetAllStudents()->front();
    std::cout << "Student: [RollNo : " << student->GetRollNo() << ", Name : " << student->GetName() << " ]"
              << std::endl;
    return 0;
}

File: Objective-C/33.TransferObjectPattern/TransferObjectPatternDemo.m
//
//  TransferObjectPatternDemo.m
//  33.TransferObjectPattern
//
//  Created by Layne on 2020/11/22.
//

#import "StudentBO.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        StudentBO* studentBusinessObject = [[StudentBO alloc] init];

        //输出所有的学生
        for (StudentVO* student in [studentBusinessObject getAllStudents]) {
            NSLog(@"Student: [RollNo : %d, Name : %@ ]", student.rollNo, student.name);
        }

        //更新学生
        StudentVO* student = [[studentBusinessObject getAllStudents] objectAtIndex:0];
        student.name = @"Michael";
        [studentBusinessObject updateStudent:student];

        //获取学生
        student = [studentBusinessObject getStudent:0];
        NSLog(@"Student: [RollNo : %d, Name : %@ ]", student.rollNo, student.name);
    }
    return 0;
}
